File: 2017/10/knot-hash-2.py
# -*- coding: utf-8 -*-
import sys
import functools
import operator


SUFFIX = [17, 31, 73, 47, 23]


def read_lengths(filename):
    with open(filename, 'rb') as f:
        content = f.read()

    return content.strip() + bytes(SUFFIX)


def build_hash(lengths, rounds=64):
    marks = list(range(256))
    skip = index = 0
    for _ in range(rounds):
        for length in lengths:
            twist(marks, index, length)
            index += length + skip
            index %= len(marks)
            skip += 1
    return marks


def reduce_hash(sparse_hash):
    dense_hash = []
    for b in range(0, 256, 16):
        block = sparse_hash[b:b + 16]
        compressed_block = functools.reduce(operator.xor, block)
        dense_hash.append(compressed_block)
    return dense_hash


def get_hex(dense_hash):
    return ''.join('{:02x}'.format(b) for b in dense_hash)


def twist(marks, index, length):
    stack = []
    for i in range(length):
        stack.append(marks[(index + i) % 256])
    for i in range(length):
        marks[(index + i) % 256] = stack.pop(-1)


def main(filename, rounds=64):
    lengths = read_lengths(filename)
    sparse_hash = build_hash(lengths, rounds=rounds)
    dense_hash = reduce_hash(sparse_hash)
    hex_hash = get_hex(dense_hash)
    print(hex_hash)


if __name__ == '__main__':
    main(*sys.argv[1:])

File: 2017/10/knot-hash-2.go
package main

import "bytes"
import "fmt"
import "io/ioutil"
import "os"


var SUFFIX = []byte{17, 31, 73, 47, 23}


type node struct {
    Value byte
    Next *node
}

type Stack struct {
    Top *node
}

func (s *Stack) Push(value byte) {
    node := &node{Value: value, Next: s.Top}
    s.Top = node
}

func (s *Stack) Pop() byte {
    node := s.Top
    s.Top = s.Top.Next
    return node.Value
}


func readLengths(filename string) []byte {
    data, err := ioutil.ReadFile(filename)
    if err != nil {
        panic(err)
    }
    return append(bytes.TrimSpace(data), SUFFIX...)
}


func buildSparseHash(lengths []byte, numRounds int) [256]byte {
    var marks [256]byte
    for i := 0; i < 256; i += 1 {
        marks[i] = byte(i)
    }
    var index byte
    var skip int
    for r := 0; r < numRounds; r += 1 {
        for _, length := range lengths {
            twist(&marks, index, length)
            index += byte(int(length) + skip)
            skip += 1
        }
    }
    return marks
}


func twist(marks *[256]byte, index byte, length byte) {
    var segment Stack
    var i byte
    for i = 0; i < length; i += 1 {
        segment.Push(marks[index + i])
    }
    for i = 0; i < length; i += 1 {
        marks[index + i] = segment.Pop()
    }
}


func reduceHash(sparseHash [256]byte) (denseHash []byte) {
    var value byte
    for i := 0; i < 256; i += 1 {
        value ^= sparseHash[i]
        if (i + 1) % 16 == 0 {
            denseHash = append(denseHash, value)
            value = 0
        }
    }
    return
}


func toHexString(hash []byte) (hexString string) {
    for _, b := range hash {
        hexString += fmt.Sprintf("%02x", b)
    }
    return
}


func main() {
    filename := os.Args[1]
    lengths := readLengths(filename)
    sparseHash := buildSparseHash(lengths, 64)
    denseHash := reduceHash(sparseHash)
    fmt.Println(toHexString(denseHash))
}
